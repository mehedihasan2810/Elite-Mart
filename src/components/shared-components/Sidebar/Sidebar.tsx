"use client";
import React, { useState } from "react";
import styles from "@/scss/shared-component-styles/Sidebar.module.scss";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "@/redux/store";
import {
  hideSidebar,
  hideSubSidebar,
  visibleSubSidebar,
} from "@/redux/createSlices/createSlice";
import Link from "next/link";

const sidebarMenu: string[] = ["Featured", "Men", "Women", "Kids", "Sale"];

const Sidebar = () => {
  const [subSidebarHeader, setSubSidebarheader] = useState<string>("");
  const dispatch = useDispatch();
  const { isSidebarVisible, isSubSidebarVisible } = useSelector(
    (state: RootState) => state.eliteMart
  );
  return (
    <aside
      onClick={(e) => {
        if (e.currentTarget !== e.target) {
          return;
        }
        dispatch(hideSidebar());
      }}
      style={{ visibility: `${isSidebarVisible ? "visible" : "hidden"}` }}
      className={styles.sidebar_bg}
    >
      <div
        style={{
          transform: `translate3d(${isSidebarVisible ? "0%" : "100%"}, 0, 0)`,
        }}
        className={styles.sidebar_container}
      >
        <div className={styles.hide_sidebar_btn_wrapper}>
          <button onClick={() => dispatch(hideSidebar())}>
            <svg
              xmlns="http://www.w3.org/2000/svg"
              fill="none"
              viewBox="0 0 24 24"
              strokeWidth={1.5}
              stroke="currentColor"
              className="w-6 h-6"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                d="M6 18L18 6M6 6l12 12"
              />
            </svg>
          </button>
        </div>

        <nav className={styles.sidebar_menu}>
          <ul role="list" className={styles.sidebar_main_nav}>
            {sidebarMenu.map((item, index) => (
              <li key={index}>
                <button
                  onClick={() => {
                    dispatch(visibleSubSidebar());
                    setSubSidebarheader(item);
                  }}
                >
                  <span>{item}</span>
                  <svg
                    xmlns="http://www.w3.org/2000/svg"
                    fill="none"
                    viewBox="0 0 24 24"
                    strokeWidth={2}
                    stroke="currentColor"
                    className="w-6 h-6"
                  >
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      d="M8.25 4.5l7.5 7.5-7.5 7.5"
                    />
                  </svg>
                </button>
              </li>
            ))}
          </ul>

          <div
            style={{
              transform: `translate3d(${
                isSubSidebarVisible ? "0%" : "100%"
              }, 0, 0)`,
            }}
            className={styles.sub_sidebar_menu}
          >
            <button onClick={() => dispatch(hideSubSidebar())}>
              {" "}
              <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={2}
                stroke="currentColor"
                className="w-6 h-6"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  d="M15.75 19.5L8.25 12l7.5-7.5"
                />
              </svg>
              <span>All</span>
            </button>
            <div className={styles.sub_sidebar_menu_wrapper}>
              <div>{subSidebarHeader}</div>
              <ul role="list">
                <li>
                  <Link
                    onClick={() => dispatch(hideSidebar())}
                    href="/products"
                  >
                    Clothing
                  </Link>
                </li>
                <li>
                  <Link
                    onClick={() => dispatch(hideSidebar())}
                    href="/products"
                  >
                    Shoes
                  </Link>
                </li>
              </ul>
            </div>
          </div>
        </nav>
      </div>
    </aside>
  );
};

export default Sidebar;
