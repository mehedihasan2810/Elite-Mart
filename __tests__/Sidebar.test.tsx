import Sidebar from "@/components/shared-components/Sidebar/Sidebar";
import { render, screen } from "@testing-library/react";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn().mockReturnValue({
    isSidebarVisible: true,
    isSubSidebarVisible: false,
  }),
}));

describe("<Sidebar />", () => {
  test("Should render properly & match snapshot", () => {
    // ARRANGE
    const { asFragment } = render(<Sidebar />);

    // ASSERT
    expect(asFragment()).toMatchSnapshot();
  });

  test("Category link should be visible", () => {
    // ARRANGE
    render(<Sidebar />);

    // ASSERT
    expect(screen.getByText("Clothing")).toBeVisible();
    ["Featured", "Men", "Women", "Kids", "Sale"].forEach((item) => {
      expect(screen.getByText(item)).toBeInTheDocument();
    });
  });
});
