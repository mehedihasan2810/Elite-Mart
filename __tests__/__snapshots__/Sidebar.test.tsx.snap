// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> Should render properly & match snapshot 1`] = `
<DocumentFragment>
  <aside
    class="sidebar_bg"
    style="visibility: visible;"
  >
    <div
      class="sidebar_container"
      style="transform: translate3d(0%, 0, 0);"
    >
      <div
        class="hide_sidebar_btn_wrapper"
      >
        <button>
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18L18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <nav
        class="sidebar_menu"
      >
        <ul
          class="sidebar_main_nav"
          role="list"
        >
          <li>
            <button>
              <span>
                Featured
              </span>
              <svg
                class="w-6 h-6"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.25 4.5l7.5 7.5-7.5 7.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </li>
          <li>
            <button>
              <span>
                Men
              </span>
              <svg
                class="w-6 h-6"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.25 4.5l7.5 7.5-7.5 7.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </li>
          <li>
            <button>
              <span>
                Women
              </span>
              <svg
                class="w-6 h-6"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.25 4.5l7.5 7.5-7.5 7.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </li>
          <li>
            <button>
              <span>
                Kids
              </span>
              <svg
                class="w-6 h-6"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.25 4.5l7.5 7.5-7.5 7.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </li>
          <li>
            <button>
              <span>
                Sale
              </span>
              <svg
                class="w-6 h-6"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.25 4.5l7.5 7.5-7.5 7.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </li>
        </ul>
        <div
          class="sub_sidebar_menu"
          style="transform: translate3d(100%, 0, 0);"
        >
          <button>
             
            <svg
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.75 19.5L8.25 12l7.5-7.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <span>
              All
            </span>
          </button>
          <div
            class="sub_sidebar_menu_wrapper"
          >
            <div />
            <ul
              role="list"
            >
              <li>
                <a
                  href="/products"
                >
                  Clothing
                </a>
              </li>
              <li>
                <a
                  href="/products"
                >
                  Shoes
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
  </aside>
</DocumentFragment>
`;
